python/tank/context.py

```diff
@@ -668,6 +668,141 @@ def create_empty(tk):
+            fallback_to_ctx_from_entity = True
+        else:
+            project = task["project"]
+            entity = task["entity"]
+            step = task["step"]
+    elif entity_type in ["PublishedFile", "TankPublishedFile"]:
+        # special case handling for published files:
+        if "task" in entity_dictionary:
+            # construct a task context
+            return from_entity_dictionary(tk, entity_dictionary["task"])
+        elif "entity" in entity_dictionary:
+            # construct an entity context
+            return from_entity_dictionary(tk, entity_dictionary["entity"])
+        elif "project" in entity_dictionary:
+            # construct project context
+            return from_entity_dictionary(tk, entity_dictionary["project"])
+        else:
+            # fall back on from_entity:
+            fallback_to_ctx_from_entity = True
+    else:
+        # find entities for an entity context
+        entity = entity_dictionary
+        if "project" not in entity:
+            fallback_to_ctx_from_entity = True
+        else:
+            project = entity["project"]
+
+    context_entities = None
+    if not fallback_to_ctx_from_entity:
+        context_entities = _populate_context_entities(project, entity, step, task)
+        fallback_to_ctx_from_entity = (context_entities is None)
+        
+    if fallback_to_ctx_from_entity:
+        # entity dict doesn't contain enough information to build a 
+        # safe, valid context so fall back on 'from_entity':
+        return from_entity(tk, entity_type, entity_id)
+
+    # update context:
+    context.update(context_entities)
+    
+    # one final check if we have a task:
+    if task:
+        additional_fields = tk.execute_core_hook("context_additional_entities").get("entity_fields_on_task", [])
+        if additional_fields:
+            # unfortunately we have to fall back to an sg query to get the additional entities :(
+            task_context = _task_from_sg(tk, task["id"], additional_fields)
+            context.update(task_context)
+    
+    return Context(**context)
+
+def _populate_context_entities(project, entity, step, task):
+    """
+    """
+    context = {}
+    
+    # check that all entity dictionaries contain id:
+    for ent in [project, entity, step, task]:
+        if ent and "id" not in ent:
+            return None
+
+    if project:
+        # add project to context:
+        project_name = project.get("name")
+        if project_name:
+            context["project"] = {"type":"Project", "id":project["id"], "name":project_name}
+        else:
+            return None
+
+    if entity:
+        # add entity to context:
+        if "type" not in entity:
+            return None
+        else:
+            name_field = {"HumanUser":"name", "Task":"content", "Project":"name"}.get(entity["type"], "code")
+            entity_name = entity.get(name_field) or entity.get("name")
```

Comment:
In the case of a Task with content being an empty string ("") this will end up with name being None rather than the empty string.  Is that an error?